### DISCUSSION TITLE
# Proposal: Allow operator extensions and more operators

### DISCUSSION THREAD

**Discussion by thargol1** [dotnet/csharplang#1941] [Language Ideas]
# Proposal: Allow operator extensions and more operators

1. Allow operators to be defined using extension methods.
2. Allow all non identifier characters and non reserved symbols to be used as operators.
3. Allow ?? as an operator.

Examples for 1. and 2.
```csharp
public static int operator Σ(this IEnumerable<int> values) 
{
	int total = 0;
	foreach(int v in values) total += v;
	return total;
}
public static double operator √(this double value) => Math.Sqrt(value);
public static bool operator ϵ(this int a, List<int> values) => values != null && values.Contains(a);
public static bool operator ¬(this bool value) => !value;
public static double operator ½(this double value) => value / 2.0;

int[] values = {1, 2, 3};
int sum = Σvalues; // sum is now 6.

double a = 4.0;
double b = √a; // b is now 2.0
double c = ½b; // c is now 1.0

var items = new List<int> {1, 2, 3};
bool check = 1 ϵ items; // check is now true
bool notCheck = ¬check; // notCheck is now false
```
Using this new operators could for example help making it easier to convert mathematical formulas to C#.

Examples for 1. and 3.
``` csharp
// result == true -> take the left part, result == false -> take the right part
public static bool operator ??(this string value) => !string.IsNullOrWhiteSpace(value);
public static bool operator ??(this TimeSpan value) => value != TimeSpan.Zero;

string a = " ";
string b = a ?? "test"; // b is now test

var c = TimeSpan.FromSeconds(0);
var d = c ?? TimeSpan.FromSeconds(1); // d is now 1 second
```

**Comment by YairHalberstadt** [comment]
These are three very different suggestions. Can I suggest you propose them all separately, in order for all of them to be given due attention?

A proposal for 1 Already exists I believe.

**Comment by bondsbw** [comment]
I personally wouldn't want to use a library that uses characters that don't exist on standard keyboards.

That doesn't mean it's a bad idea.  I could see it being useful for a project or even within a field if a virtual keyboard exists to make those characters easy to type.

**Comment by thargol1** [comment]
@YairHalberstadt yes, these are three different suggestions but suggestion 2 and 3 are dependent on suggestion 1, so splitting them up would either remove the context or introduce duplicate content.

@bondsbw I agree with you on using characters that are not on the existing keyboard. I used windows charmap to create my example. Yet the end result is shorter code and code that in some fields of work is easier to read.

**Comment by ufcpp** [comment]
1 is dup of [Shapes and Extensions](https://github.com/dotnet/csharplang/issues/164)

**Comment by ufcpp** [comment]
for 2, https://github.com/ufcpp/UfcppSample/blob/master/Demo/2018/EmojiIdentifier/swift.en.md

**Comment by Austin-bryan** [comment]
Aren't extension operators apart of "Extension Everything"?
